src/gsi/read_prepbufr.f90

```diff
@@ -689,20 +689,26 @@ subroutine read_prepbufr(nread,ndata,nodata,infile,obstype,lunout,twindin,sis,&
 
 ! identify drifting buoys - TYP=180/280 T29=562 and last three digits of SID between 500 and 999
 !  (see https://www.wmo.int/pages/prog/amp/mmop/wmo-number-rules.html)  Set kx to 199/299
-        if (id_drifter .and. (kx==180 .or. kx==280) .and. nint(hdr(3))==562) then
-           rstation_id=hdr(4)
-           read(c_station_id,*,iostat=ios) iwmo
-           if (ios == 0 .and. iwmo > 0) then
-              if(mod(iwmo,1000) >=500) then
-                 kx = kx + 19
+        ! Prevent integer overflow by nint
```

Comment:
While I see what this code doing, is this the preferred solution?   Is it acceptable for the prepbufr file to contain variables which exceed `huge_i_kind`?

Reply:
If we've added logic to safeguard against overflows, I assume we did so because this happens either in regression tests or other cases.  If true, we should document these cases and inform ObsProc.   They may be able to alter their processing to prevent problematic values from being written to prepbufr files.  Alternatively, they may identify ways in which we can improve our code to handle such cases.  My concern is that the proposed change is a patch which yields desired behavior without addressing a more fundamental issue.

Do we have an example of where this check is necessary to prevent an overflow?

Reply:
@RussTreadon-NOAA I see your point now.  I am trying to run the develop RTs on Orion to answer this question.  However, I am running into an error when global_3dvar attempts to read gsiparm.anl with iostat 18 when reading the `setup` namelist, suggesting the wrong dimensions for one or more variables is present in gsiparm.anl.  Am I interpreting that correctly?  Any suggestions on a fix?

Reply:
@DavidHuber-NOAA , I built the current head of gsi `develop` on Orion in `/work/noaa/da/rtreadon/git/gsi/develop`.   I ran ctest `global_3dvar`.  The ctests are  running OK.  

The run directory is `/work/noaa/stmp/rtreadon/develop/tmpreg_global_3dvar/`.   You'll find a `gsiparm.anl` file in any of the `global_3dvar` sub-directories.   A diff between your `gsiparm.anl` and what's in the `global_3dvar` directories may point to the issue in your `gsiparm.anl`.  

I tried to look inside `/work/noaa/nesdis-rdo2/dhuber/para/ptmp/gsi_rt/tmpreg_global_3dvar/` but can't due to permission restrictions on the directory.

Reply:
Classic mistake: I didn't load the fix submodule.

On to the debugging.  I found numerous examples of `hdr(3)` (T29) values > `huge_i_kind`.  These all seem to be the same value (100000000000.0).  I'm guessing this is some sort of fill value.  For the `/work/noaa/da/rtreadon/CASES/regtest/gfs/prod/gdas.20221109/00/atmos/gdas.t00z.prepbufr` file, these occur in records 330462 through 330520.

Reply:
I cloned [NCEPLIBS-bufr](https://github.com/NOAA-EMC/NCEPLIBS-bufr).  `BMISS` is set to 10E10, the value you found.   

When we read the prepbufr file we load data into `r_double` (real_8) arrays.   `r_double` can hold 10E10.   The problem arises when we `nint(real_8)`.    This can lead to overflow since `nint` defaults to integer_4.   Instead of adding a `if(hdr(3) < huge_i_kind)` wrapper around the code, what if we add the optional kind parameter to `nint`.   For example, instead of
```
        if(hdr(3) < huge_i_kind) then
           if (id_drifter .and. (kx==180 .or. kx==280) .and. idnint(hdr(3))==562) then
              rstation_id=hdr(4)
```
we have
```
        if (id_drifter .and. (kx==180 .or. kx==280) .and. nint(hdr(3),0_8)==562) then
              rstation_id=hdr(4)
```
Just a thought.  I am getting too involved in this review as a handling reviewer.  We need two peer reviewers to take the lead on the review.

Reply:
Switched to using `nint` with the size parameter set to `r_double`.  I left some instances of `idnint` elsewhere in the code that did not create a convoluted solution.